csrc/iter_visitor.cpp

```diff
@@ -933,4 +933,162 @@ std::vector<Val*> InputsOf::outputs(
+    ir_utils::replaceValInExpr(use, old_val, new_val);
+  }
+  return removeVal(old_val);
+}
+
+bool DeadCodeRemover::removeVal(Val* val) {
+  // Mark val dead even if we can't yet remove it due to its definition having
+  // some live outputs
+  if (!markDead(val)) {
+    // val is already marked dead
+    return false;
+  }
+
+  if (val->definition()) {
+    // If val has a definition, it can only be removed by removing its
+    // definition
+    return maybeRemoveExpr(val->definition());
+  } else {
+    TORCH_INTERNAL_ASSERT(
+        !val->isFusionInput(), "Refusing to remove Fusion input");
+    markDead(val);
+    registerRemoval(val);
+    return true;
+  }
+}
+
+bool DeadCodeRemover::maybeRemoveExpr(Expr* expr) {
+  if (allOutputsDead(expr)) {
+    if (!markDead(expr)) {
+      // Expr was already marked dead, so don't try to remove it again
+      return false;
+    }
+
+    const auto outputs = expr->outputs();
+    for (auto outp : outputs) {
+      registerRemoval(outp);
+    }
+    registerRemoval(expr);
+    return true;
+  } else {
+    return false;
+  }
+}
+
+bool DeadCodeRemover::allOutputsDead(Expr* expr) const {
+  return std::all_of(
+      expr->outputs().begin(), expr->outputs().end(), [&](Val* outp) {
+        return isDead(outp);
+      });
+}
+
+bool DeadCodeRemover::allUsesDead(Val* val) const {
+  return std::all_of(val->uses().begin(), val->uses().end(), [&](Expr* use) {
+    return isDead(use);
+  });
+}
+
+void DeadCodeRemover::markLiveRecursive(Statement* stmt) {
+  if (isLive(stmt)) {
+    return;
+  }
+  markLive(stmt);
+  if (stmt->isVal() && stmt->asVal()->definition()) {
+    markLiveRecursive(stmt);
+  } else {
+    auto expr = stmt->asExpr();
+    for (const auto inp : expr->outputs()) {
+      markLive(inp);
+    }
+    for (const auto inp : expr->inputs()) {
+      markLiveRecursive(inp);
+    }
+  }
+}
+
+bool DeadCodeRemover::markDead(Statement* stmt) {
+  if (stmt->isVal()) {
+    auto val = stmt->asVal();
+    TORCH_INTERNAL_ASSERT(
+        !val->isFusionOutput(),
+        "Call to markDead on Fusion output is illegal: ",
+        val->toString());
+    TORCH_INTERNAL_ASSERT(
+        !val->isFusionInput(),
+        "Call to markDead on Fusion input is illegal: ",
+        val->toString());
+    TORCH_INTERNAL_ASSERT(
+        allUsesDead(val),
+        "Attempted to remove Val with live uses: ",
+        val->toString());
+  }
+  return (bool)live_statements_.erase(stmt);
+}
+
+void DeadCodeRemover::doRemoval() const {
+  for (auto val : vals_to_remove_) {
+    fusion_->removeVal(val);
+  }
+  for (auto expr : exprs_to_remove_) {
+    // Fusion::removeVal(val) actually removes val->definition() from the
+    // Fusion, and sets all its outputs' definitions to nullptr. So we should
+    // not need to manually remove Exprs here. Instead, we just assert that they
+    // have already been removed if they were registered for removal.
+    TORCH_INTERNAL_ASSERT(
+        !fusion_->inContainer(expr),
```

Comment:
Is this safe? If `IrContainer::removeExpr(expr)` is already called, the memory of `expr` should have been freed at this moment, no?

Reply:
Actually `Fusion::inContainer` first checks if `expr` is in the container before dereferencing it, so it is somewhat safer than it looks. It is a little unnecessary though, since as the comment says, as long as any of its outputs were removed, the expr will be removed.

Reply:
Hmm, the code looks like using deallocated memory pointes. It's just passing them, so, yes, it should be fine...